### PR TITLE
DIRECTOR: Add support  for The Simpsons Cartoon Studio

### PR DESCRIPTION
**Pull request by donaldm**
DIRECTOR: Add support for The Simpsons Cartoon Studio

- Add msfile xlib used by The Simpsons Cartoon Studio
- Add support  for handlers that are on the stage and not the current window.
- Add earlier initialization for Score channels to fix issues with updating  immediately after a Tell (Fixes loading and saving dialog boxes)
- Fix a sizing issue with Text castmembers. In The Simpsons Cartoon Studio the open cartoon dialog was growing when changing cartoons.
- Fix an issue with Score updates reverting changed cast IDs
- Add support for filtering by a different file extension than .txt in getFileNameFromModal
- Add inflate lingo method
- Fix file encoding for FileIO::m_writeString to use platform file encoding.
- Add game quirk for The Simpsons Cartoon Studio save paths.
- Add code to update the save file cache when using getNthFileNameInFolder. This fixes an issue where if you save a cartoon and then try to load it the recently saved cartoon would not show up in the open cartoon dialog.
- Add close and forget lingo methods which are used by The Simpsons Cartoon Studio.

PS: This is my first time contributing to scummvm so please go easy on me if I have done something incorrectly in the commit process (I have tried to follow the guidelines).

### CODE REVIEW

File: common/str-enc.cpp [hunk 1]
```diff
@@ -1220,6 +1220,16 @@ U32String String::decode(CodePage page) const {
 	return unicodeString;
 }
 
+String String::encode(CodePage page) const {
```
Comment (sev-) [inline common/str-enc.cpp]: What is the purpose of this code?

Also, please put the changes to common/ as separate commits, preceding their usage in the Director engine.

File: engines/director/castmember/castmember.cpp [hunk 2]
```diff
@@ -264,6 +270,9 @@ bool CastMember::setField(int field, const Datum &d) {
 	case kTheRect:
 		warning("CastMember::setField(): Attempt to set read-only field \"%s\" of cast %d", g_lingo->field2str(field), _castId);
 		return false;
+	case kThePreLoad:
+		_preload = (bool)d.asInt();
```
Comment (sev-) [inline engines/director/castmember/castmember.cpp]: In general, we decided not to go with preloading in ScummVM. So, you may just leave it as a no-op, still processing it in these switch statements.

Could you please also provide the Lingo and how this is used in the game?

File: engines/director/game-quirks.cpp [hunk 3]
```diff
@@ -305,6 +307,10 @@ void DirectorEngine::gameQuirks(const char *target, Common::Platform platform) {
 	if (!list.empty()) {
 		CachedArchive *archive = new CachedArchive(list);
 
+		if (SearchMan.hasArchive(kQuirksCacheArchive)) {
```
Comment (sev-) [inline engines/director/game-quirks.cpp:354]: Please add a comment here why this is needed, e.g., copy/paste from the commit log message

File: engines/director/lingo/lingo-builtins.cpp [hunk 4]
```diff
@@ -121,11 +121,13 @@ static const BuiltinProto builtins[] = {
 	// Control
 	{ "abort",			LB::b_abort,		0, 0, 400, CBLTIN },	//			D4 c
 	{ "cancelIdleLoad",	LB::b_cancelIdleLoad,1, 1, 500, CBLTIN },	//				D5 c
+	{ "close",          LB::b_close,        0, 1, 400, CBLTIN },    // D2 c
```
Comment (sev-) [inline engines/director/lingo/lingo-builtins.cpp]: This and the `forget` method is actually implemented as the default methods of a `window` object. See in `lingo-object.cpp:182`. So, it seems that your problem was somewhere else, and these methods have to be removed.

File: engines/director/lingo/lingo-builtins.cpp [hunk 5]
```diff
@@ -209,6 +211,7 @@ static const BuiltinProto builtins[] = {
 	{ "updateFrame",	LB::b_updateFrame,	0, 0, 500, CBLTIN },	//				D5 c
 	// Point
 	{ "point",			LB::b_point,		2, 2, 400, FBLTIN },	//			D4 f
+	{ "inflate",        LB::b_inflate,      2, 3, 400, FBLTIN },    //			D4 f
```
Comment (sev-) [inline engines/director/lingo/lingo-builtins.cpp:214]: Is it for Point or from Rect?

The documentation says:
> This function changes the dimensions of the rectangle specified by
rectangle. The change is relative to the center of the rectangle.

File: engines/director/lingo/lingo-builtins.cpp [hunk 6]
```diff
@@ -1425,6 +1428,10 @@ void LB::b_getNthFileNameInFolder(int nargs) {
 	Datum r("");
 	Common::Array<Common::String> fileNameList;
 
+	// Update the game quirks archive in case our save state has changed.
+	// This is necessary because we may save a game and then try to open a game in the same session.
+	g_director->gameQuirks(g_director->getGameId(), g_director->getPlatform());
```
Comment (sev-) [inline engines/director/lingo/lingo-builtins.cpp:1498]: This looks like a hack or a workaround. I would need to think of a better and more generic approach here that would eplicitely refresh the relevant Archives in the SearchSet.

File: engines/director/lingo/lingo-builtins.cpp [hunk 7]
```diff
@@ -3250,6 +3307,34 @@ void LB::b_intersect(int nargs) {
 	g_lingo->push(d);
 }
 
+void LB::b_inflate(int nargs) {
+	Datum inflateHeight = g_lingo->pop();
```
Comment (sev-) [inline engines/director/lingo/lingo-builtins.cpp]:
```suggestion
	int inflateHeight = g_lingo->pop().asInt();
```

File: engines/director/lingo/lingo-builtins.cpp [hunk 8]
```diff
@@ -3250,6 +3307,34 @@ void LB::b_intersect(int nargs) {
...
+	Datum inflateWidth = g_lingo->pop();
+	Datum sR = g_lingo->pop();
+
+	TYPECHECK(inflateHeight, INT);
```
Comment (sev-) [inline engines/director/lingo/lingo-builtins.cpp]: This is not a good idea. I am sure, floats are also accepted here. It requires more investigation of the original.

File: engines/director/lingo/lingo-builtins.cpp [hunk 9]
```diff
@@ -3250,6 +3307,34 @@ void LB::b_intersect(int nargs) {
...
+	int x = sourceRect.left - inflateWidth.asInt();
+	int y = sourceRect.top - inflateHeight.asInt();
+	int width = sourceRect.width() + inflateWidth.asInt() * 2;
+	int height = sourceRect.height() + inflateHeight.asInt() * 2;
+
+	Datum inflatedRect;
+	inflatedRect.type = RECT;
+
+	inflatedRect.u.farr = new FArray();
+	inflatedRect.u.farr->arr.push_back(Datum(x));
```
Comment (sev-) [inline engines/director/lingo/lingo-builtins.cpp]:
```suggestion
	inflatedRect.u.farr->arr.push_back(sourceRect.left);
	inflatedRect.u.farr->arr.push_back(sourceRect.top);
	inflatedRect.u.farr->arr.push_back(sourceRect.right);
	inflatedRect.u.farr->arr.push_back(sourceRect.bottom);
```

File: engines/director/lingo/lingo-code.cpp [hunk 10]
```diff
@@ -1646,6 +1646,14 @@ void LC::call(const Common::String &name, int nargs, bool allowRetVal) {
 	// Handler
 	funcSym = g_lingo->getHandler(name);
 
+	bool foundInMainStage = false;
+
+	if (g_director->getCurrentMovie()->getHandler(name).type == VOIDSYM) {
```
Comment (sev-) [inline engines/director/lingo/lingo-code.cpp]: This also has to be tested with the original. We'd better create a minimal test movie demonstrating this behaviour and check it with different versions of the Director.

File: engines/director/lingo/lingo-the.cpp [hunk 11]
```diff
@@ -1730,6 +1730,16 @@ void Lingo::setTheSprite(Datum &id1, int field, Datum &d) {
 					channel->_dirty = true;
 				}
 				channel->setCast(castId);
+
+				// If we change the channel cast we also have to update the sprite in the score.
```
Comment (sev-) [inline engines/director/lingo/lingo-the.cpp]: This requires a test movie with the original

File: engines/director/lingo/xlibs/msfile.cpp [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+#include "gui/browser.h"
+#include "director/director.h"
+#include "director/picture.h"
+#include "director/types.h"
+#include "director/util.h"
+#include "director/lingo/lingo.h"
+#include "director/lingo/lingo-object.h"
+#include "director/lingo/lingo-utils.h"
+#include "director/lingo/xlibs/msfile.h"
+#include <common/file.h>
```
Comment (sev-) [inline engines/director/lingo/xlibs/msfile.cpp]:
```suggestion
#include "common/file.h"
```

File: engines/director/lingo/xlibs/msfile.cpp [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+XOBJSTUB(MSFile::m_getFullDirList, "")
+XOBJSTUB(MSFile::m_getFileList, "")
+XOBJSTUB(MSFile::m_getDirList, "")
+
+void MSFile::m_getDiskFree(int nargs) {
+	// We don't care about the drive letter. We will always just return 500 MB.
+	g_lingo->pop();
+
+	// 500 MB in KB
+	g_lingo->push(Datum(524288000));
```
Comment (sev-) [inline engines/director/lingo/xlibs/msfile.cpp]:
```suggestion
	g_lingo->push(Datum(500 * 1024 * 1024));
```
for better readability

File: engines/director/lingo/xlibs/msfile.cpp [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+}
+
+void MSFile::m_getVolLabel(int nargs) {
+	Datum d = g_lingo->pop();
+
+	Common::String volLabel = "";
```
Comment (sev-) [inline engines/director/lingo/xlibs/msfile.cpp]:
```suggestion
	Common::String volLabel;
```
It is already initialized with a zero-length string.

File: engines/director/lingo/xlibs/msfile.cpp [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+
+	const char *gameId = g_director->getGameId();
+	if (strncmp(gameId, "simpsonsstudio", 14) == 0 || strncmp(gameId, "simpsonsplayer", 14) == 0) {
+		volLabel = "Simpsons";
+	}
```
Comment (sev-) [inline engines/director/lingo/xlibs/msfile.cpp:172]:
```suggestion
	} else {
	   warning("MSFile::m_getVolLabel(): Unsupported gameid '%s'", gameId);
        }
```

File: engines/director/lingo/xlibs/msfile.cpp [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+		free(buffer);
+		delete destStream;
+		return success;
+	}
+
+	return false;
+}
+
+bool MSFile::copyGameFile(const Common::String& srcParam, const Common::String& destParam) {
+	char dirSeperator = g_director->_dirSeparator;
```
Comment (sev-) [inline engines/director/lingo/xlibs/msfile.cpp]:
```suggestion
	char dirSeparator = g_director->_dirSeparator;
```

File: engines/director/util.cpp [hunk 17]
```diff
@@ -986,6 +986,10 @@ Common::String getFileNameFromModal(bool save, const Common::String &suggested,
 	if (ext) {
 		mask += ".";
 		mask += ext;
+
+		if (strncmp(ext, "txt", 3) != 0) {
```
Comment (sev-) [inline engines/director/util.cpp:1003]: Aren't there potentially two extensions that could be added here?

File: engines/director/score.cpp [hunk 18]
```diff
@@ -1903,6 +1903,12 @@ void Score::setSpriteCasts() {
 			 j, _currentFrame->_sprites[j]->_castId.asString().c_str(), _currentFrame->_sprites[j]->_spriteType,
 			spriteType2str(_currentFrame->_sprites[j]->_spriteType));
 	}
+
+	if (_channels.size() == 0) {
```
Comment (sev-) [inline engines/director/score.cpp:1992]: I would also like to see a mninmal test movie here.

**Review by sev- — CHANGES_REQUESTED** [review]
Solid work on expanding the Director engine, thank you!

I added a few notes and questions:
  * Some slight renames
  * A lot of changes look very risky to me. In these cases we implement a small test for the original to make sure we correctly captured an edge case. If you need help, please say so
  * Please, rebase your branch to the tip of the master, we require linear history
  * Also, per our commit guidelines, please split the commit log messages into a short description (up to 80 characters, goes into the first line), followed by detailsed stuff you have now
  * Changes to common/ always must go to a separate commits
  * In general, your `encode()` method is not quite clear to me, e.g. why it should go to the common code instead of doing this enciding locally in the engine.

Looking forward to more nice stuff from you

File: engines/director/lingo/lingo-builtins.cpp (hunk 4 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: Actually the issue here is that in the lingo code the close and forget methods were being called with a window id like close(0) and forget(0). Without these methods the simpsons credit screen throws an error about the missing lingo methods.

File: engines/director/lingo/lingo-builtins.cpp (hunk 5 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp:214]: Ah this should be for rect and not point. Thanks for pointing that out!

File: engines/director/lingo/xlibs/msfile.cpp (hunk 14 above)
Comment (donaldm) [reply engines/director/lingo/xlibs/msfile.cpp]: Ah ok I can make that change, thanks for pointing that out.

File: engines/director/lingo/xlibs/msfile.cpp (hunk 15 above)
Comment (donaldm) [reply engines/director/lingo/xlibs/msfile.cpp:172]: Good idea on the else block

File: engines/director/lingo/xlibs/msfile.cpp (hunk 16 above)
Comment (donaldm) [reply engines/director/lingo/xlibs/msfile.cpp]: Good catch on the typo here, thanks!

File: engines/director/util.cpp (hunk 17 above)
Comment (donaldm) [reply engines/director/util.cpp:1003]: I am not sure I follow on this. Maybe I am not understanding something.

File: engines/director/score.cpp (hunk 18 above)
Comment (donaldm) [reply engines/director/score.cpp:1992]: I can look into that, but I am not sure how to create test movies.

File: engines/director/game-quirks.cpp (hunk 3 above)
Comment (donaldm) [reply engines/director/game-quirks.cpp:354]: I will add that, thank you

File: engines/director/castmember/castmember.cpp (hunk 2 above)
Comment (donaldm) [reply engines/director/castmember/castmember.cpp]: MissingPreloadError

I hope this screenshot shows the issue. This happens when loading a cartoon in The Simpsons Cartoon Studio. I could probably leave this as a no-op though as you suggested because it is not  set to my knowledge but Cartoon Studio checks the preLoad variable when loading.

File: common/str-enc.cpp (hunk 1 above)
Comment (donaldm) [reply common/str-enc.cpp]: This code is in conjunction with the encoding done in the FileIO xlib writeString method to fix cartoon saves in the cartoon studio. Without this the file encoding was incorrect and it didn't handle unprintable characters so the cartoons could not  be loaded.

I put it where I did so I could make it similar to the existing decode method. It seemed like a cleaner approach to me.

Comment (donaldm) [reply common/str-enc.cpp]: I will break this off into its own commit

File: engines/director/lingo/lingo-builtins.cpp (hunk 4 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: CreditsCloseHandlerMissing_Pic2
CreditsForgetMethodMissing_Pic3
Here are relevant screenshots

File: engines/director/lingo/lingo-builtins.cpp (hunk 6 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp:1498]: I can understand that. I couldn't think of a better way to implement this and was just trying to make freshly saved cartoons show up in the loading cartoon dialog. If there was a way to have some sort of file watch or something so that when the saves get changed the cache would get refreshed that might be a cleaner approach (I just didn't feel I had the skills to implement that).

File: engines/director/lingo/lingo-builtins.cpp (hunk 7 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: Ah ok I can do that, I just didn't see this style being done in other code that I saw. This is a simple enough change.

File: engines/director/lingo/lingo-builtins.cpp (hunk 8 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: Should I remove the TYPECHECK? It is probably not needed because the code should handle either.

File: engines/director/lingo/lingo-builtins.cpp (hunk 9 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: I don't think this change would work because x and y are computed values to expand the rectangle around the center. I tested this with Simpsons Cartoon Studio and comparing the result to what I saw running the game under winevdm.

File: engines/director/lingo/lingo-code.cpp (hunk 10 above)
Comment (donaldm) [reply engines/director/lingo/lingo-code.cpp]: I am actually not sure how to do this. I based these changes from debugging Cartoon Studio. I could use some help here.

File: engines/director/lingo/lingo-the.cpp (hunk 11 above)
Comment (donaldm) [reply engines/director/lingo/lingo-the.cpp]: I could use some help creating these test movies because I am not sure how to do that.

File: engines/director/lingo/xlibs/msfile.cpp (hunk 12 above)
Comment (donaldm) [reply engines/director/lingo/xlibs/msfile.cpp]: Ah thanks, I can fix this typo.

File: engines/director/lingo/xlibs/msfile.cpp (hunk 13 above)
Comment (donaldm) [reply engines/director/lingo/xlibs/msfile.cpp]: Good idea, I can do that.

File: engines/director/util.cpp (hunk 17 above)
Comment (sev-) [reply engines/director/util.cpp:1003]: Consider `ext` to be `dat`. We first add `.dat`, then your check is passing and you add `.txt` on top. E.g. resulting in `file.dat.txt`

Comment (donaldm) [reply engines/director/util.cpp:1003]: Yes, I did that deliberately because otherwise files wouldn't be displayed in the file selection dialog due to the way ScummVM does save files. A lot of the code seems to require the extension to end in .txt to work, but I wanted to be able to "filter" on ctd files which are used by The Simpsons Cartoon Studio and not all ScummVM save files.

Comment (donaldm) [reply engines/director/util.cpp:1003]: Here is a screenshot showing what I am talking about:

importdialog

File: engines/director/lingo/lingo-builtins.cpp (hunk 6 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp:1498]: One thing I will say is when I looked it appeared that this was the only place this cache was being used. When I search the entire project for kQuirksCacheArchive I only find it actually used in the LB::b_getNthFileNameInFolder method which is why I took this approach.

File: common/str-enc.cpp (hunk 1 above)
Comment (donaldm) [reply common/str-enc.cpp]: I have moved this into its own commit prefixed by COMMON.

File: engines/director/castmember/castmember.cpp (hunk 2 above)
Comment (donaldm) [reply engines/director/castmember/castmember.cpp]: I made this a no-op and removed the preLoad field. I added a warning when setting the preLoad field to inform that ScummVM does not handle preloading.

File: engines/director/lingo/lingo-builtins.cpp (hunk 5 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp:214]: I fixed the comment here. The other methods were also rect methods.

File: engines/director/lingo/lingo-builtins.cpp (hunk 7 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: I reworked the code here.

File: engines/director/lingo/lingo-builtins.cpp (hunk 8 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: I removed this typecheck.

File: engines/director/lingo/xlibs/msfile.cpp (hunk 15 above)
Comment (donaldm) [reply engines/director/lingo/xlibs/msfile.cpp:172]: I added this else statement.

File: engines/director/lingo/xlibs/msfile.cpp (hunk 16 above)
Comment (donaldm) [reply engines/director/lingo/xlibs/msfile.cpp]: Fixed.

File: engines/director/game-quirks.cpp (hunk 3 above)
Comment (donaldm) [reply engines/director/game-quirks.cpp:354]: Done.

File: common/str.h [hunk 19]
```diff
@@ -173,6 +173,8 @@ class String : public BaseString<char> {
 	/** Python-like method **/
 	U32String decode(CodePage page = kUtf8) const;
 
+	String encode(CodePage page = kUtf8) const;
```
Comment (ccawley2011) [inline common/str.h]: This should have two parameters - one for the source code page, and one for the destination code page. Currently, the source code page is hardcoded to UTF-8.

File: engines/director/lingo/xlibs/msfile.cpp [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+		bool error = srcStream->err();
+		delete srcStream;
+
+		if (error) {
+			free(buffer);
+			return false;
+		}
+
+		// Write the source file buffer to the destination file. This is effectively a file copy.
+		destStream->write(buffer, size);
```
Comment (ccawley2011) [inline engines/director/lingo/xlibs/msfile.cpp]: You could use `destStream->writeStream(srcStream);` to simplify the code and avoid creating larger buffers in memory.

File: common/str.h (hunk 19 above)
Comment (donaldm) [reply common/str.h]: Thank you for the code review. I can certainly add a second parameter, but I was trying to make it similar to the existing decode function My understanding is that internally ScummVM uses a UTF-8 encoding for it's strings so I didn't think it was necessary. Is that understanding incorrect? If it is should we do a similar approach for the decode method as well?

File: engines/director/lingo/xlibs/msfile.cpp (hunk 20 above)
Comment (donaldm) [reply engines/director/lingo/xlibs/msfile.cpp]: Thank you for the advice, I can look into implementing that. I can see how that would be more efficient especially for large files!

File: engines/director/lingo/lingo-code.cpp (hunk 10 above)
Comment (moralrecordings) [reply engines/director/lingo/lingo-code.cpp]: Have a look at https://github.com/scummvm/director-tests/tree/master/D4-unit - essentially we're after some Director files which replicate the exact situation this modification fixes, and therefore fails with the original code and passes with the new code. So in this case it would do something like the following:
- stage (movie A) opens child window (movie B)
- movie B attempts to call a function that only exists in movie A
- control (and the context, so cast, sprites etc.) is confirmed to have switched to movie A

As sev mentioned, we have to be very precise about mods being made to integral bits of the engine like the call resolver; Director games love to rely on every possible edge case. That said windowing support definitely needs improving and I'm excited that someone is having a stab at it!

File: common/str.h (hunk 19 above)
Comment (ccawley2011) [reply common/str.h]: The encoding used by `String` is unspecified, but it's often used for non-UTF8 strings. The parameter for  `String::decode()` refers to the destination code page used by the `String` (while the source code page for `U32String` is effectively always UTF-32/UCS-4).

If you specify the source code page as the last parameter, that should help with keeping `String::encode()` and `U32String::encode()` consistent with each other.

File: engines/director/lingo/lingo-builtins.cpp (hunk 4 above)
Comment (sev-) [reply engines/director/lingo/lingo-builtins.cpp]:
```suggestion
	{ "close",          LB::b_close,        0, 1, 400, CBLTIN },    //			D4 c
```

These were added in D4, not in D2

File: engines/director/lingo/lingo-builtins.cpp [hunk 21]
```diff
@@ -121,11 +121,13 @@ static const BuiltinProto builtins[] = {
 	// Control
 	{ "abort",			LB::b_abort,		0, 0, 400, CBLTIN },	//			D4 c
 	{ "cancelIdleLoad",	LB::b_cancelIdleLoad,1, 1, 500, CBLTIN },	//				D5 c
+	{ "close",          LB::b_close,        0, 1, 400, CBLTIN },    // D2 c
 	{ "continue",		LB::b_continue,		0, 0, 200, CBLTIN },	// D2 c
 	{ "dontPassEvent",	LB::b_dontPassEvent,0, 0, 200, CBLTIN },	// D2 c
 	{ "delay",	 		LB::b_delay,		1, 1, 200, CBLTIN },	// D2 c
 	{ "do",		 		LB::b_do,			1, 1, 200, CBLTIN },	// D2 c
 	{ "finishIdleLoad",	LB::b_finishIdleLoad,1, 1, 500, CBLTIN },	//				D5 c
+	{ "forget",         LB::b_forget,       0, 1, 200, CBLTIN },    // D2 c
```
Comment (sev-) [inline engines/director/lingo/lingo-builtins.cpp]:
```suggestion
	{ "forget",         LB::b_forget,       0, 1, 400, CBLTIN },    //			D4 c
```

File: engines/director/lingo/lingo-builtins.cpp [hunk 22]
```diff
@@ -1425,6 +1429,10 @@ void LB::b_getNthFileNameInFolder(int nargs) {
 	Datum r("");
 	Common::Array<Common::String> fileNameList;
 
+	// Update the game quirks archive in case our save state has changed.
```
Comment (sev-) [inline engines/director/lingo/lingo-builtins.cpp:1496]: What does "in case our save state has changed" mean? I do not see any related checks, the update is unconditional here.

File: engines/director/lingo/lingo-builtins.cpp (hunk 4 above)
Comment (sev-) [reply engines/director/lingo/lingo-builtins.cpp]: The described syntax in the documentation is the following:
Screenshot 2025-07-21 at 12 14 44

Does it still work? You may add lingo tests, and how to run the tests is described at https://wiki.scummvm.org/index.php?title=Director#Executing_lingo_tests

It is important to clarify because we do not want to break other titles by this change.

If you need help with testing, please come to our Discord and we will give a helping hand, there is #engine-director channel

File: engines/director/lingo/lingo-builtins.cpp [hunk 23]
```diff
@@ -1669,6 +1695,38 @@ void LB::b_finishIdleLoad(int nargs) {
...
+			if (movie)
+				movie->getScore()->_playState = kPlayStopped;
+
+			// remove me from global vars
+			for (auto &it : g_lingo->_globalvars) {
+				if (it._value.type != OBJECT || it._value.u.obj->getObjType() != kWindowObj)
+					continue;
+
+				if (it._value.u.obj == window)
+					g_lingo->_globalvars[it._key] = 0;
```
Comment (sev-) [inline engines/director/lingo/lingo-builtins.cpp]: This does not look like removing. Should not this call `erase()` method?

Also, this seems to contradict to the description of this command:
Screenshot 2025-07-21 at 12 21 36
Please denote part "when the window is no longer in use and no other variables refer to it"

In fact, this better should be checked with the original, on what is happening when:
* Window is open
* Window is open but invisible
* Some variable(s) have reference to this window

If you need help with checking this behaviour, please say so

File: engines/director/lingo/lingo-builtins.cpp [hunk 24]
```diff
@@ -3250,6 +3308,35 @@ void LB::b_intersect(int nargs) {
...
+	TYPECHECK(sR, RECT);
+
+	Common::Rect sourceRect(sR.u.farr->arr[0].asInt(),
+		                    sR.u.farr->arr[1].asInt(),
+		                    sR.u.farr->arr[2].asInt(),
+		                    sR.u.farr->arr[3].asInt());
+
+	int inflatedLeft = sourceRect.left - inflateWidth;
+	int inflatedTop = sourceRect.top - inflateHeight;
+	int inflatedRight = inflatedLeft + sourceRect.width() + inflateWidth * 2;
```
Comment (sev-) [inline engines/director/lingo/lingo-builtins.cpp]:
```suggestion
	int inflatedRight = sourceRect.right + inflateWidth;
```
It is much easier to read, and mathematically it is equivalent. Same for the line below.

File: engines/director/lingo/lingo-the.cpp (hunk 11 above)
Comment (sev-) [reply engines/director/lingo/lingo-the.cpp]: Please join our Discord, ping me, and I will explain and give you all the needed tools

File: engines/director/lingo/lingo-the.cpp [hunk 25]
```diff
@@ -2441,6 +2451,8 @@ void Lingo::setObjectProp(Datum &obj, Common::String &propName, Datum &val) {
 		} else {
 			g_lingo->lingoError("Lingo::setObjectProp: Object <%s> has no property '%s'", obj.asString(true).c_str(), propName.c_str());
 		}
+	} else if (obj.type == INT) {
```
Comment (sev-) [inline engines/director/lingo/lingo-the.cpp]: What is the usecase that you saw?

File: common/str.h (hunk 19 above)
Comment (donaldm) [reply common/str.h]: Thanks for the clarification, I will work to add that.

File: engines/director/lingo/lingo-builtins.cpp (hunk 21 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: Thank you and no worries about the time it takes to respond, I took a few days off myself. I will make this correction

File: engines/director/lingo/lingo-builtins.cpp (hunk 22 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp:1496]: Right it is unconditional. The cache is originally only created at the start of the game and then never invalidated. We call it here so that when this method is called the cache is rebuilt in case a save file has been created at any point during the game. It would probably be better to invalidate the cache when a save file is created/saved but as far as I am aware there could be a variety of ways that a save file could be created. I wasn't sure how to implement this better in ScummVM. If there are alternative ideas to invalidate/refresh the cache here I am open to suggestions.

File: engines/director/lingo/lingo-builtins.cpp (hunk 23 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: I based this code on the existing forget command, but if I misunderstood something here I can look into this further. I will research this more.

File: engines/director/lingo/lingo-builtins.cpp (hunk 24 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: Thanks for the suggestion, I will work to make this change and test the result to make sure it functions as expected.

File: engines/director/lingo/lingo-the.cpp (hunk 25 above)
Comment (donaldm) [reply engines/director/lingo/lingo-the.cpp]: I saw this throw an error in Simpsons Cartoon Studio, but I will find the exact code involved for this.

**Comment by donaldm** [comment]
> Apologies for a slow review, I am on a holiday these days.
> 
> Please, join our Discord at https://discord.gg/4cDsMNtcpG and ping me. I will explain how we test with the original and create test movies.

Thanks, I will work to join the Discord for this so I can learn how to create these test movies.

File: engines/director/lingo/lingo-builtins.cpp (hunk 24 above)
Comment (donaldm) [reply engines/director/lingo/lingo-builtins.cpp]: Fixed.

File: engines/director/lingo/lingo-the.cpp (hunk 25 above)
Comment (donaldm) [reply engines/director/lingo/lingo-the.cpp]: CreditsInvalidObject0

File: engines/director/lingo/xlibs/msfile.cpp (hunk 20 above)
Comment (donaldm) [reply engines/director/lingo/xlibs/msfile.cpp]: Fixed.

File: engines/director/score.cpp (hunk 18 above)
Comment (donaldm) [reply engines/director/score.cpp:1992]: I will work to create a test movie for this

File: engines/director/lingo/lingo-the.cpp (hunk 25 above)
Comment (rvanlaar) [reply engines/director/lingo/lingo-the.cpp]: When showing logs you use `--debugflags=lingoexec,lingostrict -d5` This way we can see what leads up to the lingo error.

**Comment by donaldm** [comment]
I met with Sev yesterday and we worked through a number of the outstanding issues for this PR. Based on the review with him I removed two of the most controversial commits from this because we determined that they were no longer needed (the addition of the close/forget methods, and the handler on a different stage issue which we determined is not needed currently. 

I am currently in the process of determining a new fix for the score issue to fix the cartoon studio dialogs and making several other changes based on feedback.

File: engines/director/score.cpp (hunk 18 above)
Comment (donaldm) [reply engines/director/score.cpp:1992]: Confirmed with Sev that this was a good change.

**Comment by sev-** [comment]
Thank you!
